datalad/support/network.py

```diff
@@ -969,38 +1007,43 @@ def is_ssh(ri):
         or (isinstance(_ri, URL) and _ri.scheme == 'ssh')
 
 
-def get_local_file_url(fname, compatibility='git-annex'):
+def get_local_file_url(fname: str,
+                       compatibility: str = 'git-annex',
+                       allow_relative_path: bool = True
+                       ) -> str:
     """Return OS specific URL pointing to a local file
 
     Parameters
     ----------
     fname : string
         Filename.  If not absolute, abspath is used
-    compatibility : {'git', 'git-annex'}, optional
-        On Windows, and only on that platform, file:// URLs may need to look
-        different depending on the use case or consuming application. This
-        switch selects different compatibility modes: 'git' for use with
-        Git commands (e.g. `clone` or `submodule add`); 'git-annex` for
-        Git-annex command input (e.g. `addurl`). On any other platform this
-        setting has no effect.
+    compatibility : str, optional
+        This parameter is only interpreted on Windows systems. If set to
+        anything else than 'git', the anchor, e.g. `C:` of `fname` will be put
+        into the `file-auth` part, i.e. network location, defined in RFC 8089.
+        This option is mainly used to support git-annex specific encoding of
+        Windows paths.
+    allow_relative_path: bool, optional
+        Allow `fname` to be a relative path. The path will be converted to an
+        absolute path, by using the current directory as path prefix.
     """
-    # we resolve the path to circumwent potential short paths on unfortunate
-    # platforms that would ruin the URL format.
-    #path = Path(fname).resolve().absolute()
-    path = Path(fname).resolve()
-    if on_windows:
-        # in addition we need to absolute(), as on windows resolve() doesn't
-        # imply that
-        path = path.absolute().as_posix()
-        furl = 'file://{}{}'.format(
-            '/' if compatibility == 'git' else '',
-            urlquote(
-                re.sub(r'([a-zA-Z]):', r'\1', path)
-            ))
-    else:
-        # TODO:  need to fix for all the encoding etc
-        furl = str(URL(scheme='file', path=str(path)))
-    return furl
+    path_obj = Path(fname)
+    if not path_obj.is_absolute():
+        if allow_relative_path is True:
+            fname = str(Path(os.getcwd()) / path_obj)
+        else:
+            raise ValueError('cannot create file-URL with relative path')
+
+    url_path = local_path2url_path(fname, auto_resolve=False)
+    if on_windows and compatibility != "git":
+        # Work around the way in which git-annex interprets file URLs on
+        # Windows. This code path will put the path anchor, e.g. `C:` of `fname`
+        # into the network location component of the resulting URL.
+        return "file:/" + url_path
+
+    result = "file://" + url_path
+    assert result == Path(fname).as_uri()
```

Comment:
how strong believe in this here?  I would leave such assertion for tests, and here if really desired -- just added a warning if mismatch detected.  Do we want entire process to crash if this assertion is not fulfilled? (e.g. some trailing `/` gets in etc)

Reply:
I am quite confident, and I think it might be good to blow instead of continuing with a faulty URI, which might -by chance- lead to data loss. Also, an uncaught exception might expose problems in our code earlier.

But I will change it to a warning,

Done in a194ecc5195db1e636866fe6825e765a3f2bf1b9